### PR TITLE
fix(TokenHoldersPanel): Added sorting feature

### LINKED ISSUE #11032: Remote destruct popup: switch to `SortableTokenHoldersList`
# Bug Report

## Description

Currently `TokenHoldersPanel` is used which doesn't support sorting. To align it with the current design sorting should be supported.

Warning: Figma designs in progress

https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?type=design&node-id=30716-634011&t=Rw4edL9LegNa9ez6-0

Screenshot from 2023-06-12 13-42-41

### PR DESCRIPTION
**Pull request by alexandraB99**
Closes #11032

### What does the PR do
TokenHoldersPanel: Added sorting feature

### Affected areas
TokenHoldersPanel

### Screenshot of functionality (including design for comparison)

https://github.com/status-im/status-desktop/assets/31625338/54a50d92-2d9b-432e-8c71-5122356a6f77

https://github.com/status-im/status-desktop/assets/31625338/386cd482-3170-4678-a7bb-c08b0c42dde5

### CODE REVIEW

**Comment by alexandraB99** [comment]
@benjthayer could you please have a look at this, we're wondering if we need sorting in the destroy column. Programmatically is a bit complex and not straight forward to do.

One good point from @micieslak "it's all implementable ofc, proxy model could be added with additional role, but sorting on the "destroy" value is a bit problematic - e.g. when it's enabled (soring on destroy column) changing value in dropdown would cause instant reordering"

Could you please have a look and let us know if this is needed and what should happen in the above case and in general all use cases?

cc @noeliaSD

File: ui/app/AppLayouts/Communities/panels/SortableTokenHoldersList.qml [hunk 1]
```diff
@@ -54,9 +46,9 @@ StatusListView {
             d.resetOtherHeaders(this)
 
             if (sorting === StatusSortableColumnHeader.Sorting.Ascending)
-                d.sorting = SortableTokenHoldersList.Sorting.Ascending
+                d.sorting = TokenHoldersProxyModel.Sorting.Ascending
```
Comment (caybro) [inline ui/app/AppLayouts/Communities/panels/SortableTokenHoldersList.qml:49]: What about using the generic `Qt::SortOrder` enum?

File: ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+            }
+        }
+
+        onClicked: {
+            d.resetOtherHeaders(this)
+
+            if (sorting === StatusSortableColumnHeader.Sorting.Ascending)
+                d.sorting = TokenHoldersProxyModel.Sorting.Ascending
+            else if (sorting === StatusSortableColumnHeader.Sorting.Descending)
+                d.sorting = TokenHoldersProxyModel.Sorting.Descending
```
Comment (caybro) [inline ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml:49]: Here as well, you could directly reuse `Qt.DescendingOrder` and `Qt.AscendingOrder`

**Review by caybro — APPROVED** [review]
LGTM

**Comment by John-44** [comment]
> > One good point from @micieslak "it's all implementable ofc, proxy model could be added with additional role, but sorting on the "destroy" value is a bit problematic - e.g. when it's enabled (soring on destroy column) changing value in dropdown would cause instant reordering"
> 
> This is a good point. I think it would be fine to remove the sorting from the destroy col - when we know more about how people will use this dialog and what use cases we want to support, we could look at enhancing the interaction design as needed. WDYT @John-44 ?

If sorting on the destroy column is difficult to do I would leave it for now.  We have many other important things to get built, so if this will take significant additional effort let's drop it for the time being.  We can always come back to it later.

Thanks!

**Review by noeliaSD — CHANGES_REQUESTED** [review]
Tested only in `storybook`. One issue found:

- The checker value is not kept neither the selected combo value when sorting is done. See the video for better understanding:

https://github.com/status-im/status-desktop/assets/97019400/ef233226-f2d2-4872-ad26-1a4423ecad41

**Review by micieslak — CHANGES_REQUESTED** [review]
It's strictly related to what @noeliaSD noticed. The state (number of tokens to destroy and selection) is kept within delegate. Delegates are dynamically created and removed when going out of viewport and the state is lost causing improper behavior observed by @noeliaSD .

File: ui/app/AppLayouts/Communities/panels/TokenHoldersProxyModel.qml [hunk 3]
```diff
@@ -6,9 +6,13 @@ SortFilterProxyModel {
     property string searchText
     readonly property string searchTextLowerCase: searchText.toLowerCase()
 
-    property int sortBy: SortableTokenHoldersList.SortBy.Username
+    property int sortBy: TokenHoldersProxyModel.SortBy.Username
     property int sortOrder: Qt.AscendingOrder
 
+    enum SortBy {
```
Comment (micieslak) [inline ui/app/AppLayouts/Communities/panels/TokenHoldersProxyModel.qml:12]: I like it, it's definitely better place for that enum :)

**Comment by alexandraB99** [comment]
> It's strictly related to what @noeliaSD noticed. The state (number of tokens to destroy and selection) is kept within delegate. Delegates are dynamically created and removed when going out of viewport and the state is lost causing improper behavior observed by @noeliaSD .

also when sorting the model resets, so the dropdown can't know what was chosen. what do you suggest?

**Comment by alexandraB99** [comment]
@noeliaSD @micieslak all set, you may wanna have a second look

File: storybook/pages/SortableTokenHoldersListPage.qml [hunk 4]
```diff
@@ -31,7 +31,7 @@ SplitView {
                 sourceModel: tokenHoldersModel
 
                 sortBy: holdersList.sortBy
-                sortOrder: holdersList.sorting === SortableTokenHoldersList.Sorting.Descending
+                sortOrder: holdersList.sorting === Qt.DescendingOrder
```
Comment (caybro) [inline storybook/pages/SortableTokenHoldersListPage.qml:34]:
```suggestion
                sortOrder: holdersList.sorting
```

File: ui/app/AppLayouts/Communities/panels/SortableTokenHoldersList.qml [hunk 5]
```diff
@@ -73,8 +65,8 @@ StatusListView {
     QtObject {
         id: d
 
-        property int sortBy: SortableTokenHoldersList.SortBy.None
-        property int sorting: SortableTokenHoldersList.Sorting.Descending
+        property int sortBy: TokenHoldersProxyModel.SortBy.None
+        property int sorting: Qt.DescendingOrder
```
Comment (caybro) [inline ui/app/AppLayouts/Communities/panels/SortableTokenHoldersList.qml:69]: What about just `sortOrder`, to match https://doc.qt.io/qt-5/qt.html#SortOrder-enum

File: ui/app/AppLayouts/Communities/panels/TokenHoldersPanel.qml [hunk 6]
```diff
@@ -1,164 +1,100 @@
...
-    QtObject {
-        id: d
+    TokenHoldersProxyModel {
+        id: filteredModel
+        sourceModel: root.model
+        searchText: searcher.text
 
-        readonly property int red2Color: 4
+        sortBy: holdersList.sortBy
+        sortOrder: holdersList.sorting === Qt.DescendingOrder
```
Comment (caybro) [inline ui/app/AppLayouts/Communities/panels/TokenHoldersPanel.qml:32]:
```suggestion
        sortOrder: holdersList.sorting
```

Comment (caybro) [reply ui/app/AppLayouts/Communities/panels/TokenHoldersPanel.qml:32]: Would look nicer with `holdersList.sortOrder` ;)

File: ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                leftPadding: 0
+                rightPadding: 0
+                sensor.enabled: false
+                title: formattedTitle
+                statusListItemTitle.visible: !unknownHolder
+                subTitle: Utils.getElidedPk(model.walletAddress)
+                asset.name: model.imageSource
+                asset.isImage: true
+                asset.isLetterIdenticon: !asset.name
+                asset.color: Theme.palette.userCustomizationColors[d.red2Color]
```
Comment (caybro) [inline ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml:188]:
```suggestion
                asset.color: Theme.palette.getColor("red2")
```

File: ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    }
+
+    QtObject {
+        id: d
+        property int sortBy: TokenHoldersProxyModel.SortBy.None
+        property int sorting: Qt.DescendingOrder
+        property var selectedTokenAmount: new Map();
+        property var selectedTokenChecked: new Map();
+
+        readonly property int red2Color: 4
```
Comment (caybro) [inline ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml:70]: You can remove this magic number; it's the `red2` named color from `StatusColors.colors`

**Review by noeliaSD — COMMENTED** [review]
Reviewed again!! Thanks for the fixes it works better now!!  :)

However I still see 2 issues here:
1. The button amount summary is not correctly calculated after sorting.
2. There is something wrong with the visualization of the complete list since depending on the sorting option there are some holder members that disappear (if you see the video, in some point `carmen` disappears).

See video for better understanding:

https://github.com/status-im/status-desktop/assets/97019400/84436fb4-7091-4214-8adc-1275db30ce23

File: ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml (hunk 2 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml]: done

File: storybook/pages/SortableTokenHoldersListPage.qml (hunk 4 above)
Comment (alexandraB99) [reply storybook/pages/SortableTokenHoldersListPage.qml]: done

File: ui/app/AppLayouts/Communities/panels/SortableTokenHoldersList.qml (hunk 5 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Communities/panels/SortableTokenHoldersList.qml:69]: done

File: ui/app/AppLayouts/Communities/panels/TokenHoldersPanel.qml (hunk 6 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Communities/panels/TokenHoldersPanel.qml]: done

File: ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml (hunk 7 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml]: done

File: ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml (hunk 8 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml]: done

**Comment by alexandraB99** [comment]
> Reviewed again!! Thanks for the fixes it works better now!! :)
> 
> However I still see 2 issues here:
> 
>     1. The button amount summary is not correctly calculated after sorting.
> 
>     2. There is something wrong with the visualization of the complete list since depending on the sorting option there are some holder members that disappear (if you see the video, in some point `carmen` disappears).
> 
> 
> See video for better understanding:
> Screen.Recording.2023-07-21.at.18.26.20.mov

All fixed now

**Review by micieslak — COMMENTED** [review]
Looks nice in general, but there are some differences in comparison to the design regarding scrolling behavior:

Screenshot from 2023-07-24 18-11-45

File: ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+                onCheckStateChanged: {
+                    if (checked)
+                        root.selfDestructAmountChanged(model.walletAddress, d.selectedTokenAmount.get(walletAddress))
+                    else
+                        root.selfDestructRemoved(model.walletAddress)
+
+                    d.selectedTokenChecked.set(walletAddress, checked);
+                }
+                Component.onCompleted: {
+                    checked = !!d.selectedTokenChecked.get(walletAddress) ? d.selectedTokenChecked.get(walletAddress) : false;
```
[the remark this answers is not in the document]
Comment (alexandraB99) [reply ui/app/AppLayouts/Communities/panels/TokenHoldersList.qml]: done

File: ui/app/AppLayouts/Communities/panels/TokenHoldersPanel.qml [hunk 10]
```diff
@@ -1,164 +1,98 @@
...
 import shared.controls 1.0
 
 Control {
     id: root
-
-    // Expected roles: name, walletAddress, imageSource, amount
+    //394 by design
+    implicitHeight: (column.implicitHeight < 394 ?
+                    column.implicitHeight : 394) +
+                    Style.current.padding
```
[the remark this answers is not in the document]
Comment (alexandraB99) [reply ui/app/AppLayouts/Communities/panels/TokenHoldersPanel.qml]: done

**Review by micieslak — CHANGES_REQUESTED** [review]
I haven't seen the code but the look and behavior needs adjustments to fulfill the design:

1. The gray think line should be semi-transparent (4 pixels). Only one most external 1-px line is opaque:

Design:
Screenshot from 2023-07-27 10-55-31
Screenshot from 2023-07-27 10-56-34

Impl:

Screenshot from 2023-07-27 11-02-39


2. The bottom thick gray line should be visible whenever the list in other than bottom-most position. In the bottom most position it's thin gray line:

Design:
Screenshot from 2023-07-27 11-04-07
Screenshot from 2023-07-27 11-04-21

Impl:
Screenshot from 2023-07-27 11-05-17

Screenshot from 2023-07-27 11-05-12

3. Scrolling down should first hide search box:
Design:
image

Impl:
Screenshot from 2023-07-27 11-09-43

4. Spacing for item after separator is invalid:
Screenshot from 2023-07-27 10-52-01
Screenshot from 2023-07-27 10-52-04

5. Top bar has bigger height than in the design:
Design:
Screenshot from 2023-07-27 11-16-37

Impl:
Screenshot 2023-07-27 11:17:42

6. Spacing of right columns looks substantially different:
Design:
Screenshot from 2023-07-27 11-20-39

Impl:

Screenshot from 2023-07-27 11-21-38

7. There are some clipping problems visible when the window is small:

Screenshot from 2023-07-27 11-23-07

8. The initial value of combo box should be equal to holding:
Design:
Screenshot from 2023-07-27 11-28-19
Impl:

Screenshot from 2023-07-27 11-29-16

9. When the list is short, popup height should adapt to avoid empty spaces:
Impl:
image

10. There is no `[Token name] token holders` label:
Design:
Screenshot from 2023-07-27 11-36-21
Impl:

Screenshot from 2023-07-27 11-36-25

**Review by micieslak — CHANGES_REQUESTED** [review]
Not sure what went wrong, but sizing seems to be broken. Console output is clear, but behavior is as on the video:

https://github.com/status-im/status-desktop/assets/20650004/a9a16934-e97a-411f-8fde-d12d2930836c

webm version:

[Screencast 2023-07-28 10_42_34.webm](https://github.com/status-im/status-desktop/assets/20650004/001c2274-32b2-4bf0-b28c-020910c1e2ae)

**Comment by alexandraB99** [comment]
@micieslak issues 2,3,4,7,8,9,10 should be fixed. please check once again

**Review by micieslak — CHANGES_REQUESTED** [review]
First, search box cannot be hidden immediately. It needs to be scrollable to work properly. Otherwise there are problems with jumpy scrollbar, sizing and the behavior is not intuitive.

Here, using mouse scroll on top of the popup:

[Screencast 2023-07-28 17_09_50.webm](https://github.com/status-im/status-desktop/assets/20650004/d089bd8c-0add-45fb-9201-9253162e6067)

Issue 4 is done partially, it the design separator occupies additional space thanks to that items seem to distributed evenly:

Screenshot 2023-07-28 17:38:12
Screenshot 2023-07-28 17:38:39


Issue 10 seems to be stil there

I also noticed some strange delegate misplacement when searching:

[Screencast 2023-07-28 17_21_35.webm](https://github.com/status-im/status-desktop/assets/20650004/5e905c00-af1d-4157-9c80-b719b7d19c10)

**Review by micieslak — COMMENTED** [review]
There is still problem with delegate's layout (holdings text, combo box and checkbox compressed in the right side) and misplacement when using filtering. When layout gets broken after searching it's impossible to select/deselect the checkbox which overlaps with scrollbar. Imo it requires some confirmation if it happens always or maybe only in my env.

[Kazam_screencast_00000.webm](https://github.com/status-im/status-desktop/assets/20650004/28d91254-2c62-442f-9dcf-e56c9471e084)

When searching, according to the design, there should be no label "Search result":

Impl:
Screenshot from 2023-08-03 00-35-12

Design:
Screenshot from 2023-08-03 01-19-26

Another (minor) issue is spacing regarding separator when sorting by username (mentioned in previous comment) and missing lack of "Cancel" button in the footer.

**Review by micieslak — COMMENTED** [review]
Currently it's impossible to scroll to the very end, so last entry is unreachable. 

[Kazam_screencast_00004.webm](https://github.com/status-im/status-desktop/assets/20650004/a7846a24-938a-4071-a95c-8351d3e04b03)

**Review by micieslak — APPROVED** [review]
There is still some misbehavior regarding bottom scrolling edge. As it's not mission-critical I'm approving :)

[Kazam_screencast_00005.webm](https://github.com/status-im/status-desktop/assets/20650004/ca07ac7c-8af4-4d5d-8c36-346639473f60)

**Pull request merged**